servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/DefaultNettyConnection.java

```diff
@@ -408,7 +409,8 @@ private boolean failIfWriteActive(ChannelOutboundListener newChannelOutboundList
             // It is possible that we have set the writeSubscriber, then the channel becomes inactive, and we will
             // never notify the write writeSubscriber of the inactive event. So if the channel is inactive we notify
             // the writeSubscriber.
-            if (!channel().isActive()) {
+            // It is also possible that Channel is in closing state, so we need to prevent sending more requests.
+            if (!channel().isActive() || closeReason != null) {
```

Comment:
`closeReason` isn't `volatile` so this may never see the `closeReason`.

Instead of using `closeReason` here use the new pipeline event to replace `writableListener` with a "poison value" that indicates abort writes and also aborts the current write if present.

Reply:
Good catch!
I have to introduce another volatile boolean to prevent new writes because:
1. Can not reuse `AbortWritesEvent` because it's emitted only when there are additional requests in the pipeline. If there is no pipelining and all requests are sequential, it's not emitted => does not abort new writes.
2. Tried to reuse `ProtocolPayloadEndEvent.OUTBOUND` but it has a much broader scope (also used for h2) and fails a lot of tests.
3. Handling of "poison value" for `channelOutboundListener` complicates internal state management. It adds at least 2 volatile reads (1st in `cleanupOnWriteTerminated, 2nd in `failIfWriteActive`) and complicates `isWriteActive` implementation.

Reply:
Discussed offline. Because `closeReason` is not on the hot path it's better just to make `closeReason` volatile instead of expanding internal state with another boolean.